Understand welkin (#1015)
elastisys/welkin: docs/operator-manual/understand-welkin.md

For a basic understanding of Welkin from the point-of-view of the Application Developers, head to [Application Developer Overview](../user-guide/index.md).

## Welkin Architecture

To begin with, you should familiarize yourself with [Welkin's architecture](../architecture.md).
It describes what components are part of Welkin and what component talks to which other components.
It allows you to create a mental model and reason about complex failure modes, such as "a buffer overflow in fluentd may be caused by OpenSearch lacking sufficient capacity to ingest all logs".

In particular, notice that:

- Welkin is composed of at least two Kubernetes clusters:
  - at least one Workload Cluster: this hosts the application(s) of the Application Developer; and
  - one Service Cluster: this provides several [Service Endpoints](../glossary.md#service-endpoint) to the Application Developers, in particular around authentication and observability.
- Welkin is composed of two layers:
  - The Kubernetes-lifecycle layer sets up rather vanilla Kubernetes Clusters which some security defaults. This layer is implemented either via Kubespray or Cluster API.

>   - The Welkin Apps layer augments the two Kubernetes Clusters with projects around security and observability. This layers has a single implementation.

Reviewer: ```suggestion
  - The Welkin Apps layer augments the two Kubernetes Clusters with projects around security and observability. This layer has a single implementation.
```

Author: Fixed, thanks!

Cluster API offers features such as Cluster autoscaling, Node self-healing and faster Kubernetes upgrades.
On the downside, it requires a mature Cluster API provider.
Not all infrastructure providers have a mature Cluster API provider.

Therefore, Welkin Environments can also be set up using Kubespray.
Kubespray runs on pretty much any infrastructure you throw at it.
However, it lacks support for Cluster autoscaling and Kubernetes upgrades are slow (albeit reliable).

For both Cluster API and Kubespray, the infrastructure provider might come with its own limitations, such as lack of load-balancers and/or of block storage which integrate with Kubernetes.
The [provider audit](provider-audit.md) is a systematic way to discover and understand these differences, so as to configure Welkin properly.

## Configuration Repository

Welkin abides to the configuration-as-code and infrastructure-as-code principles.
In essence, all Welkin configuration is stored in text files (most frequently YAML).
Welkin expects you to stores these files in a git repository, which is why we refer to Welkin configuration as configuration repository.

> A typical configuration repository looks as follows:

Reviewer: Since the config repos looks quite different for Kubespray and CAPI it might be good to specify that this is the example structure for when you are using [Welkin Kubespray](https://github.com/elastisys/compliantkubernetes-kubespray):
```suggestion
A typical configuration repository when using Kubespray looks as follows:
```

- `adminlogs` is a folder where you can put text files (Markdown) describing operations you performed against this environment.
This folder is optional and ignored by Welkin.
We recommend having separate folders for changes, incidents and maintenance.
- `.sops.yaml` contains [sops](https://github.com/getsops/sops) configuration, used by Welkin to encrypt secrets.
Welkin makes sure to decrypt secrets when it needs them.

The Welkin Kubespray layer initializes and reads the following configuration files and folders:

- `sc-config` and `wc-config` store configuration for the Service Cluster and Workload Cluster, respectively.
These folders are consumed by Ansible, which is part of this layer.
If `group_vars`, `all` and `inventory.ini` look new to you, we recommend you learn more about Ansible.

The Welkin Apps layer initializes and reads the following configuration files and folders:

- `backups` is a folder in which Welkin stores copies of previous configurations.
See this as a convenience, given that all configuration is already in git.

> - `common-config.yaml` contains configuration common to both the Service Cluster and the Workload Cluster.
> - `wc-config.yaml` and `sc-config.yaml` contains configuration specific to the Service Cluster and the Workload Cluster, respectively.

Reviewer: Maybe clarify that these config files _overrides_ the config found in the `defaults` folder:
```suggestion
- `common-config.yaml` contains override configuration common to both the Service Cluster and the Workload Cluster.
- `wc-config.yaml` and `sc-config.yaml` contains override configuration specific to the Service Cluster and the Workload Cluster, respectively.
```

For a basic understanding of Welkin from the point-of-view of the Application Developers, head to [Application Developer Overview](../user-guide/index.md).

## Welkin Architecture

To begin with, you should familiarize yourself with [Welkin's architecture](../architecture.md).
It describes what components are part of Welkin and what component talks to which other components.
It allows you to create a mental model and reason about complex failure modes, such as "a buffer overflow in Fluentd may be caused by OpenSearch lacking sufficient capacity to ingest all logs".

In particular, notice that:

- Welkin is composed of at least two Kubernetes clusters:
    - at least one Workload Cluster: this hosts the application(s) of the Application Developer; and
    - one Service Cluster: this provides several [Service Endpoints](../glossary.md#service-endpoint) to the Application Developers, in particular around authentication and observability.
- Welkin is composed of two layers:
    - The Kubernetes-lifecycle layer sets up rather vanilla Kubernetes Clusters with some security defaults. This layer is implemented either via Kubespray or Cluster API.

>     - The Welkin Apps layer augments the two Kubernetes Clusters with projects around security and observability. This layer has a single implementation.

Reviewer 2: Maybe should be some note here about Management cluster when using CAPI? noting that the SC cluster becomes the MC and manages both itself and WC. But this could also be a bit confusing, let me know what you think.

Author: Good point! Fixed.

## Kubespray vs Cluster API

Although Welkin tries to be a platform which is as portable as possible, the two implementations of the Kubernetes-lifecycle layer differ quite a bit.
As a Platform Administrator, you need to be aware of these differences.

Welkin recommends using the Cluster API implementation when setting up a new Environment.
Cluster API offers features such as Cluster autoscaling, Node self-healing and faster Kubernetes upgrades.
On the downside, it requires a mature Cluster API provider.
Not all infrastructure providers have a mature Cluster API provider.

Therefore, Welkin Environments can also be set up using Kubespray.
Kubespray runs on pretty much any infrastructure you throw at it.

> However, it lacks support for Cluster autoscaling and Kubernetes upgrades are slow (albeit reliable).

Reviewer 2: ```suggestion
However, it lacks support for Cluster autoscaling and Kubernetes upgrades are slower (albeit reliable).
```
slow feels subjective. I'd recommend using slower in comparison to CAPI.

For both Cluster API and Kubespray, the infrastructure provider might come with its own limitations, such as lack of load-balancers and/or of block storage which integrate with Kubernetes.
The [provider audit](provider-audit.md) is a systematic way to discover and understand these differences, so as to configure Welkin properly.

## Configuration Repository

Welkin abides to the configuration-as-code and infrastructure-as-code principles.
In essence, all Welkin configuration is stored in text files (most frequently YAML).
Welkin expects you to stores these files in a git repository, which is why we refer to Welkin configuration as configuration repository.

> A typical configuration repository when using [Welkin Kubespray](https://github.com/elastisys/compliantkubernetes-kubespray) looks as follows:

Reviewer 2: Do we want to use the word "repository" here? It's more of the structure of the configuration, but it might be confusing because it doesn't have to be a git repository for example.

Author: I agree that "repository" is a bit "overkill", given that we don't technically require the configuration to be in a git repository. However, I'd like to strengthen the point that Welkin is GitOps-ish, so I'd keep the term like this for now.

> A typical configuration repository when using [Welkin Kubespray](https://github.com/elastisys/compliantkubernetes-kubespray) looks as follows:

Reviewer 3: If Cluster API is the recommended way, I'd appreciate **that** to be demonstrated here. Or some words around why Kubespray is used here as an example. 

It is not hard to argue why Kubespray is used: it is after all the most generic and most "supportive" one, given that it can run on basically any Linux server.

Author: Good point! I added a description of files consumed by CAPI too.

```console
my-welkin-environment
|-- adminlog
|   `-- changes
|   `-- incidents
|   `-- maintenance
|-- backups
|-- common-config.yaml
|-- compliantkubernetes-apps
|-- compliantkubernetes-kubespray
|-- defaults
|   |-- common-config.yaml
|   |-- sc-config.yaml
|   `-- wc-config.yaml
|-- README.md
|-- sc-config
|   |-- group_vars
|   |   |-- all
|   |   |   |-- ck8s-kubespray-general.yaml
|   |   |   `-- ck8s-ssh-keys.yaml
|   |   |-- etcd
|   |   |   `-- ck8s-etcd.yaml
|   |   `-- k8s_cluster
|   |       |-- ck8s-k8s-cluster-default.yaml
|   |       `-- ck8s-k8s-cluster.yaml
|   `-- inventory.ini
|-- sc-config.yaml
|-- secrets.yaml
|-- .sops.yaml
|-- .state
|   |-- kube_config_sc.yaml
|   |-- kube_config_wc.yaml
|   `-- s3cfg.ini
|-- wc-config
|   |-- group_vars
|   |   |-- all
|   |   |   |-- ck8s-kubespray-general.yaml
|   |   |   `-- ck8s-ssh-keys.yaml
|   |   |-- etcd
|   |   |   `-- ck8s-etcd.yaml
|   |   `-- k8s_cluster
|   |       |-- ck8s-k8s-cluster-default.yaml
|   |       `-- ck8s-k8s-cluster.yaml
|   `-- inventory.ini
`-- wc-config.yaml

> ```

Reviewer 2: Just wanna add that this is a very creative way of showing the structure of the config, I like it 😄

Author: Thanks! Unfortunately, I cannot claim originality. I just copied from the best out there. :smile:

Let us dive into the role of each of these files in the order you would commonly look at them:

- `README.md` is a human description of the environment.
This file is optional and ignored by Welkin.
We recommend describing here the propose of the Environment, what kind of applications it hosts, what infrastructure provider it runs on and any noteworthy deviations.
- `adminlogs` is a folder where you can put text files (Markdown) describing operations you performed against this environment.
This folder is optional and ignored by Welkin.
We recommend having separate folders for changes, incidents and maintenance.
- `.sops.yaml` contains [sops](https://github.com/getsops/sops) configuration, used by Welkin to encrypt secrets.

> Welkin makes sure to decrypt secrets when it needs them.

Reviewer 3: I think it would be good to just add a single sentence about where the scripts that do this are, because they are not shown in the file listing.

It should perhaps be written that all the installation and configurations are done by scripts, and that the configuration is stored in the repository. Specifically calling out the GitOps-friendliness of all this is good, too, because that's a keyword people may be looking for.

Author: Indeed, we should call out "GitOps-ness". Done!

The Welkin Kubespray layer initializes and reads the following configuration files and folders:

- `sc-config` and `wc-config` store configuration for the Service Cluster and Workload Cluster, respectively.
These folders are consumed by Ansible, which is part of this layer.

> If `group_vars`, `all` and `inventory.ini` look new to you, we recommend you learn more about Ansible.

Reviewer 2: Could maybe add a link to where they could learn more about Ansible?

Author: True, we can point them back to "Understand the Basics", i.e., a soft-RTFM. :smile: Fixed with thanks!

The Welkin Apps layer initializes and reads the following configuration files and folders:

> - `backups` is a folder in which Welkin stores copies of previous configurations.

Reviewer 2: Worth mentioning backup retention time here, or maybe information overload? 😄

Author: Information overload, but it does make sense to clarify that these configuration backups are unrelated to application and platform data backups.

Instead, override configuration values with the files described below.
- `common-config.yaml` contains override configuration common to both the Service Cluster and the Workload Cluster.
- `wc-config.yaml` and `sc-config.yaml` contains override configuration specific to the Service Cluster and the Workload Cluster, respectively.
- `secrets.yaml` contains secrets, both for the Workload Cluster and Service Cluster.
This file is encrypted using the information in `.sops.yaml`.
You should only edit this file using sops to make sure secrets never end up in plain text in the configuration repository.

Except for `init` sub-commands and configuration migration steps, the files above are only consumed by Welkin.
They only take as input explicit choices that you made and describe how an Environment should be.

The following files and folders are stored by Welkin in the configuration repository and contain information from an Environment:

- `.state` is a folder which contains the "state", i.e., files produced by Welkin needed to access an Environment during daily operations, in particular kubeconfig and S3 object storage access.

> If needed, these files will be encrypted with sops.

Reviewer 3: ```suggestion
Typically, these files should be encrypted with sops for information security purposes.
```

In Welkin, kubeconfigs generally don't contain credentials, only pointer to OpenID configuration, hence are usually not encrypted.

To make sure your whole team knows exactly which version of Welkin an Environment runs, it is common practice to add the source code, in the example above `compliantkubernetes-apps` and `compliantkubernetes-kubespray`, as git submodules.

> An alternative is to pack all Welkin source code in a Docker image.

Reviewer 3: Is this described somewhere or will it be?

Author: Unfortunately not. We might write one later, but this is out-of-scope for the present PR.

> 

Reviewer 3: If we're going with Kubespray as an example, would it be possible to state which files have the interesting settings in a Cluster API configuration as well? I mean the configuration parameters for e.g. kubeadm and such.

Author: Done!

## What does it mean to use Welkin as a Platform Administrator on a daily basis?

You must specify a configuration repository in the environment variable `CK8S_CONFIG_PATH`.
Then you can interact with Welkin via commands such as `ck8s` and `ck8s-kubespray` provided in the source code.

> A typical usage of these commands is describe later in this guide.

Reviewer 2: I'm guessing this guide is a WIP? Or does it just end here.
```suggestion
A typical usage of these commands are described later in this guide.
```

Reviewer 3: The "it" refers to the singular "usage", so it's fine. :)

Author: I added a section "Next Steps" to clarify that this is the end of this page. :smile:
